src/code/z_fireobj.c

```diff
@@ -1,27 +1,244 @@
+    u8 nextState;
+    Vec3f dist;
+
+    FireObj_StepSize(fire);
+    fire->unk26 += 1;
+    if (fire->ignitionDelay > 0) {
+        fire->ignitionDelay--;
+    } else if (fire->ignitionDelay == 0) {
+        fire->ignitionDelay = -1;
+        if ((fire->state == 3) || ((fire->state == 1))) {
+            nextState = 0;
+        } else {
+            nextState = 1;
+        }
+        FireObj_SetState(fire, fire->dynamicSizeStep, nextState);
+    }
+    if (((fire->flags & 1) != 0) && (fire->state != 3) &&
+        (func_800CA1E8(globalCtx, &globalCtx->colCtx, fire->position.x, fire->position.z, &sp44, &waterBox) != 0) &&
+        ((fire->yScale * (0, 6500.0f)) < (sp44 - fire->position.y))) { // Fake but IDK what else
+        FireObj_SetState(fire, fire->dynamicSizeStep, 3);
+    }
+    if (((fire->flags & 2) != 0) && (player->itemActionParam == 7)) {
+        Math_Vec3f_Diff(&player->swordInfo[0].tip, &fire->position, &dist);
+        if (Math3D_LengthSquared(&dist) < 400.0f) {
+            sp40 = 1;
+        }
+    }
+    if (sp40 != 0) {
+        if (fire->state == 3) {
+            if (player->unk_B28 > 0) {
+                FireObj_SetState(fire, fire->dynamicSizeStep, 0);
+            }
+        } else if (player->unk_B28 == 0) {
+            player->unk_B28 = 0xD2;
+            Audio_PlaySoundAtPosition(globalCtx, &fire->position, 20, NA_SE_EV_FLAME_IGNITION);
+        } else if (player->unk_B28 < 0xC8) {
+            player->unk_B28 = 0xC8;
+        }
+    }
+}
+
+void FireObj_Draw(GlobalContext* globalCtx, FireObj* fire) {
+    s32 pad;
+    FireObjLightParams* temp_v1 = &sFireObjLightParams[fire->unk27];
+
+    if (fire->state != 3) {
+        Vec3s vec;
+        OPEN_DISPS(globalCtx->state.gfxCtx);
+        func_8012C2DC(globalCtx->state.gfxCtx);
+        gSPSegment(POLY_XLU_DISP++, 0x08,
+                   Gfx_TwoTexScroll(globalCtx->state.gfxCtx, 0, 0, 0, 0x20, 0x40, 1, 0, (fire->unk26 * -0x14) & 0x1FF,
+                                    0x20, 0x80));
+
+        gDPSetPrimColor(POLY_XLU_DISP++, 0, temp_v1->unk4, temp_v1->unk0, temp_v1->unk1, temp_v1->unk2, temp_v1->unk3);
+
+        gDPSetEnvColor(POLY_XLU_DISP++, temp_v1->unk5, temp_v1->unk6, temp_v1->unk7, 0);
+
+        vec.x = 0;
+        vec.y = func_800DFCDC(globalCtx->cameraPtrs[globalCtx->activeCamera]) + 0x8000;
+        vec.z = 0;
+        Matrix_SetStateRotationAndTranslation(fire->position.x, fire->position.y, fire->position.z, &vec);
+        Matrix_Scale(fire->xScale, fire->yScale, 1.0f, 1);
+
+        gSPMatrix(POLY_XLU_DISP++, Matrix_NewMtx(globalCtx->state.gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+        gSPDisplayList(POLY_XLU_DISP++, D_0407D590);
+        CLOSE_DISPS(globalCtx->state.gfxCtx);
+    }
+}
+
+#ifdef NON_EQUIVALENT
+// Accesses data incorrectly
+void FireObj_InitLight(GlobalContext* globalCtx, FireObjLight* light, u8* param_3, Vec3f* pos) {
+    FireObjLightParams* objectParams;
+
+    objectParams = &sFireObjLightParams[*param_3];
```

Comment:
imo the code is cleaner this way but I don't think we have any particular preference, so this is optional.
```suggestion
    FireObjLightParams* objectParams = &sFireObjLightParams[*param_3];
```